##
## Copyright (c) 2016-2019 Genomics England Ltd.
##
## This file is part of PanelApp
## (see https://panelapp.genomicsengland.co.uk).
##
## Licensed to the Apache Software Foundation (ASF) under one
## or more contributor license agreements.  See the NOTICE file
## distributed with this work for additional information
## regarding copyright ownership.  The ASF licenses this file
## to you under the Apache License, Version 2.0 (the
## "License"); you may not use this file except in compliance
## with the License.  You may obtain a copy of the License at
##
##   http://www.apache.org/licenses/LICENSE-2.0
##
## Unless required by applicable law or agreed to in writing,
## software distributed under the License is distributed on an
## "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
## KIND, either express or implied.  See the License for the
## specific language governing permissions and limitations
## under the License.
##
# Generated by Django 2.0.6 on 2018-07-04 15:47

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        ('panels', '0059_merge_20180704_1224'),
    ]

    operations = [
        migrations.RemoveField(
            model_name='genepanelsnapshot',
            name='current_number_of_evaluated_regions',
        ),
        migrations.RemoveField(
            model_name='genepanelsnapshot',
            name='current_number_of_regions',
        ),
        migrations.RemoveField(
            model_name='region',
            name='type_of_effects',
        ),
        migrations.RemoveField(
            model_name='region',
            name='type_of_variants',
        ),
        migrations.AddField(
            model_name='region',
            name='haploinsufficiency_score',
            field=models.CharField(choices=[('3', 'Sufficient evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('2', 'Emerging evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('1', 'Little evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('0', 'No evidence to suggest that dosage sensitivity is associated with clinical phenotype'), ('40', 'Dosage sensitivity unlikely'), ('30', 'Gene associated with autosomal recessive phenotype')], default='0', max_length=2),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='region',
            name='triplosensitivity_score',
            field=models.CharField(choices=[('3', 'Sufficient evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('2', 'Emerging evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('1', 'Little evidence suggesting dosage sensitivity is associated with clinical phenotype'), ('0', 'No evidence to suggest that dosage sensitivity is associated with clinical phenotype'), ('40', 'Dosage sensitivity unlikely'), ('30', 'Gene associated with autosomal recessive phenotype')], default='1', max_length=2),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='region',
            name='verbose_name',
            field=models.CharField(blank=True, help_text='ISCA Region Name', max_length=256, null=True),
        ),
        migrations.AlterField(
            model_name='region',
            name='gene_core',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.PROTECT, to='panels.Gene'),
        ),
        migrations.AlterField(
            model_name='region',
            name='mode_of_pathogenicity',
            field=models.CharField(blank=True, choices=[('', 'Provide exceptions to loss-of-function'), ('Loss-of-function variants (as defined in pop up message) DO NOT cause this phenotype - please provide details in the comments', 'Loss-of-function variants (as defined in pop up message) DO NOT cause this phenotype - please provide details in the comments'), ('Other', 'Other - please provide details in the comments')], max_length=255, null=True),
        ),
        migrations.AlterField(
            model_name='region',
            name='moi',
            field=models.CharField(choices=[('', 'Provide a mode of inheritance'), ('MONOALLELIC, autosomal or pseudoautosomal, NOT imprinted', 'MONOALLELIC, autosomal or pseudoautosomal, NOT imprinted'), ('MONOALLELIC, autosomal or pseudoautosomal, maternally imprinted (paternal allele expressed)', 'MONOALLELIC, autosomal or pseudoautosomal, maternally imprinted (paternal allele expressed)'), ('MONOALLELIC, autosomal or pseudoautosomal, paternally imprinted (maternal allele expressed)', 'MONOALLELIC, autosomal or pseudoautosomal, paternally imprinted (maternal allele expressed)'), ('MONOALLELIC, autosomal or pseudoautosomal, imprinted status unknown', 'MONOALLELIC, autosomal or pseudoautosomal, imprinted status unknown'), ('BIALLELIC, autosomal or pseudoautosomal', 'BIALLELIC, autosomal or pseudoautosomal'), ('BOTH monoallelic and biallelic, autosomal or pseudoautosomal', 'BOTH monoallelic and biallelic, autosomal or pseudoautosomal'), ('BOTH monoallelic and biallelic (but BIALLELIC mutations cause a more SEVERE disease form), autosomal or pseudoautosomal', 'BOTH monoallelic and biallelic (but BIALLELIC mutations cause a more SEVERE disease form), autosomal or pseudoautosomal'), ('X-LINKED: hemizygous mutation in males, biallelic mutations in females', 'X-LINKED: hemizygous mutation in males, biallelic mutations in females'), ('X-LINKED: hemizygous mutation in males, monoallelic mutations in females may cause disease (may be less severe, later onset than males)', 'X-LINKED: hemizygous mutation in males, monoallelic mutations in females may cause disease (may be less severe, later onset than males)'), ('MITOCHONDRIAL', 'MITOCHONDRIAL'), ('Unknown', 'Unknown'), ('Other', 'Other - please specifiy in evaluation comments')], max_length=255, verbose_name='Mode of inheritance'),
        ),
        migrations.AlterField(
            model_name='region',
            name='name',
            field=models.CharField(help_text='ISCA ID', max_length=128),
        ),
        migrations.AlterField(
            model_name='region',
            name='panel',
            field=models.ForeignKey(on_delete=django.db.models.deletion.PROTECT, to='panels.GenePanelSnapshot'),
        ),
        migrations.AlterField(
            model_name='trackrecord',
            name='issue_type',
            field=models.CharField(choices=[('Created', 'Created'), ('NewSource', 'Added New Source'), ('RemovedSource', 'Removed Source'), ('ChangedGeneName', 'Changed Gene Name'), ('SetPhenotypes', 'Set Phenotypes'), ('SetModelofInheritance', 'Set Mode of Inheritance'), ('ClearSources', 'Clear Sources'), ('SetModeofPathogenicity', 'Set mode of pathogenicity'), ('GeneClassifiedbyGenomicsEnglandCurator', 'Gene classified by Genomics England curator'), ('EntityClassifiedbyGenomicsEnglandCurator', 'Entity classified by Genomics England curator'), ('SetModeofInheritance', 'Set mode of inheritance'), ('SetPenetrance', 'Set penetrance'), ('SetPublications', 'Set publications'), ('ApprovedGene', 'Approved Gene'), ('ApprovedEntity', 'Approved Entity'), ('GelStatusUpdate', 'Gel Status Update'), ('UploadGeneInformation', 'Upload gene information'), ('RemovedTag', 'Removed Tag'), ('AddedTag', 'Added Tag'), ('ChangedName', 'Changed Name'), ('ChangedSTRName', 'Changed STR Name'), ('ChangedChromosome', 'Changed Chromosome'), ('ChangedPosition37', 'Changed GRCh37'), ('ChangedPosition38', 'Changed GRCh38'), ('ChangedNormalRepeats', 'Changed Normal Number of Repeats'), ('ChangedPathogenicRepeats', 'Changed Pathogenic Number of Repeats'), ('RemovedGene', 'Removed Gene from the STR'), ('ChangedRepeatedSequence', 'Changed Repeated Sequence'), ('ChangedEffectTypes', 'Changed Effect Types'), ('ChangedVariantType', 'Changed Variant Types')], max_length=512),
        ),
    ]
